### PR TITLE
Initial checkin - ensembl vep and lookup

### PR DESCRIPTION
**Pull request by bwalsh**
This PR contains 1 new feature:
* docker container exposing REST interface for ensembl vep and lookup services
* volume testing shows approximately 4 sec response time.

Known issues:
* due to slow build times, docker build broken into two steps: ensembl-base contains perl and data , ensemble-rest configures endpoint
* first request to each endpoint is slow due to startup
* local mysql data has not been setup ftp://ftp.ensembl.org/pub/grch37/release-92/mysql/

--README--

# ensembl-rest
Language agnostic RESTful data access to Ensembl data over HTTP
Inspired by andrewyatz's vagrant effort https://github.com/andrewyatz/vagrant_machines/tree/master/ensembl/rest
## pre-requisites
docker
## setup perl environment, ubuntu packages, etc.
docker build -t ensembl-base -f Dockerfile-base .
## configure it
docker build -t ensembl-rest -f Dockerfile-rest  .
## start it
docker run -d  --name ensembl-rest -p 3000:3000  -it ensembl-rest
## test
curl 'http://localhost:3000/vep/human/hgvs/ENSP00000361021.3:p.Arg15Lys?domains=1&protein=1&uniprot=1'  -H "Accept: application/json"  | jq '.[0].id == "ENSP00000361021.3:p.Arg15Lys"'
curl 'http://localhost:3000/lookup/id/ENST00000275493?expand=1'  -H 'Content-type:application/json' | jq '.id == "ENST00000275493"'
### installation
https://github.com/Ensembl/ensembl-rest/wiki/REST-Installation-and-Development
### api
https://rest.ensembl.org/documentation/info/lookup
https://rest.ensembl.org/documentation/info/vep_hgvs_get

### CODE REVIEW

File: Dockerfile-base [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ADD settings/ ./settings
+
+
+ENV ENSEMBL_API_VERSION ${ENSEMBL_API_VERSION:-92} 
+
+# Set TEST_ENV to nothing to stop deploying a MariaDB along with your server
+ENV TEST_ENV true
+  # config.vm.network :forwarded_port, guest: 3000, host: 3000
+  # config.vm.network :forwarded_port, guest: 5000, host: 5000
+  # config.vm.synced_folder "rest-src/", "/home/vagrant/src", :create => true
```
Comment (buchanae) [inline Dockerfile-base]: What are these for?

File: Dockerfile-base [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+# ENV TABIX_HOME=/root/src/tabix-master
+# ENV PERL5LIB=${PERL5LIB}:${TABIX_HOME}/perl/blib/lib:${TABIX_HOME}/perl/blib/arch
+
+# # Setup LD_LIBRARY_PATH
+# ENV LD_LIBRARY_PATH=${TABIX_HOME}
+
+# WORKDIR /root/src/ensembl-rest
+# ADD ensembl-rest.conf /root/src/ensembl-rest/ensembl-rest.conf
+# CMD ./script/ensembl_rest_server.pl
```
Comment (buchanae) [inline Dockerfile-base]: Are you planning to uncomment this at some point?

File: Dockerfile-rest [hunk 3]
```diff
@@ -0,0 +1,26 @@
+FROM ensembl-base
```
Comment (buchanae) [inline Dockerfile-rest:1]: Why separate the base from the second half?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,22 @@
...
+Language agnostic RESTful data access to Ensembl data over HTTP
+Inspired by andrewyatz's vagrant effort https://github.com/andrewyatz/vagrant_machines/tree/master/ensembl/rest
+## pre-requisites
+docker
+## setup perl environment, ubuntu packages, etc.
+docker build -t ensembl-base -f Dockerfile-base . 
+## configure it
+docker build -t ensembl-rest -f Dockerfile-rest  . 
+## start it 
+docker run -d  --name ensembl-rest -p 3000:3000  -it ensembl-rest
```
Comment (buchanae) [inline README.md:11]: Should we publish this to ohsucompbio/ensemble-rest?

File: Dockerfile-base [hunk 5]
```diff
@@ -0,0 +1,30 @@
+FROM ubuntu
+ADD deploy/ ./deploy
+ADD settings/ ./settings
+
+
+ENV ENSEMBL_API_VERSION ${ENSEMBL_API_VERSION:-92} 
+
+# Set TEST_ENV to nothing to stop deploying a MariaDB along with your server
+ENV TEST_ENV true
```
Comment (buchanae) [inline Dockerfile-base]: I don't see any volumes defined. Where is the mysql database living? Will it be destroyed across restarts?

File: deploy/deploy.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  echo "mysql-server mysql-server/root_password_again select vagrant" | debconf-set-selections
+  apt-get install -y mariadb-server
+
+  service mysql stop
+  cp /settings/my.cnf /etc/mysql/my.cnf
+  service mysql start
+
+  /usr/bin/mysql -uroot -pvagrant -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' IDENTIFIED BY 'vagrant' WITH GRANT OPTION; FLUSH PRIVILEGES"
+  /usr/bin/mysql -uroot -pvagrant -e "CREATE USER 'ro'@'%'; GRANT SELECT ON *.* TO 'ro'@'%'; FLUSH PRIVILEGES"
+  /usr/bin/mysql -uroot -pvagrant -e "CREATE USER 'rw'@'%'; GRANT ALL PRIVILEGES ON *.* TO 'rw'@'%'; FLUSH PRIVILEGES"
```
Comment (buchanae) [inline deploy/deploy.sh]: Probably want to include this user/pass config detail in the readme.

File: deploy/setup.sh [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+cd /root/src/ensembl-rest
+cpanm --installdeps .
+
+# Confirming the setup
+perl Makefile.PL
+
+
+wget ftp://ftp.ensembl.org/pub/release-75/fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.75.dna.toplevel.fa.gz -O /root/.vep/Homo_sapiens.GRCh37.75.dna.toplevel.fa.gz
+cd /root/.vep
+uncompress Homo_sapiens.GRCh37.75.dna.toplevel.fa.gz
```
Comment (buchanae) [inline deploy/setup.sh:45]: This is a large data file, and I expect there will be others. What's your plan for those? Include them in the docker image? Or require the user to download them and provide instructions? Or other?

File: ensembl_rest.conf [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+copyright_footer = <a href='http://www.sanger.ac.uk'>WTSI</a> &amp; <a href='http://www.ebi.ac.uk'>EBI</a>
+
+jsonp=1
+
+<Model::Registry>
+  ###### Database settings. Use if you want to connect to a single database instance. Common options are given below
+
+  host = ensembldb.ensembl.org
+  port = 3337
+  user = anonymous
```
Comment (buchanae) [inline ensembl_rest.conf:22]: What's going on here? Are we still connecting to an external service in some cases?

**Review by buchanae — APPROVED** [review]
A few minor comments and questions, but looks good.

File: Dockerfile-base (hunk 1 above)
Comment (bwalsh) [reply Dockerfile-base]: Good catch.  I'll delete

File: Dockerfile-base (hunk 2 above)
Comment (bwalsh) [reply Dockerfile-base]: Good catch. I'll delete

File: Dockerfile-rest (hunk 3 above)
Comment (bwalsh) [reply Dockerfile-rest:1]: The main reason was build times,  it takes ~ 1hr to build the base.   By separating into two steps, we can deploy different, relatively lightweight rest endpoints

File: README.md (hunk 4 above)
Comment (bwalsh) [reply README.md:11]: Sounds like a plan

File: deploy/deploy.sh (hunk 6 above)
Comment (bwalsh) [reply deploy/deploy.sh]: agreed.

File: Dockerfile-base (hunk 5 above)
Comment (bwalsh) [reply Dockerfile-base]: I'll discuss in known issues

File: ensembl_rest.conf (hunk 8 above)
Comment (bwalsh) [reply ensembl_rest.conf:22]: yes, the lookup service  uses this external db ( I commented on this in known issues)

**Pull request merged**